mllib/src/main/scala/org/apache/spark/mllib/stat/test/ChiSqTest.scala

```diff
@@ -195,7 +200,15 @@ private[spark] object ChiSqTest extends Logging {
       }
     }
     val df = size - 1
-    val pValue = 1.0 - new ChiSquaredDistribution(df).cumulativeProbability(statistic)
+    val pValue = if (simulatePValue && !expArr.isEmpty) {
+      val spark = SparkSession.getActiveSession.getOrElse(SparkSession.getDefaultSession.get)
+      val exp: BDV[Double] = BDV(expArr.map(_ * scale))
+      val digest = getChi2Digest(spark, exp, numDraw = numDraw)
+
+      1.0 - digest.cdf(statistic)
+    } else {
+      1.0 - new ChiSquaredDistribution(df).cumulativeProbability(statistic)
+    }
```

Comment:
What's the intuition of doing MC given we can compute p-value directly?

Reply:
For small N (<5-10 depending on who you talk to) in any bucket of the array of expected values, the theoretical chi2 distribution is not valid so we cannot use it in the goodness of fit test. In such cases, we can use MC to empirically compute the distribution of the chi2 metric, and use this distribution in lieu of the theoretical one. 

Here's an image depicting the deviation of this empirical distribution from the theoretical one as we scan the total number of points (note that this number in the legend is not the "N in any bucket" I referred to above, but think of it as a proxy).
![image](https://user-images.githubusercontent.com/4906224/68055001-efd81100-fcac-11e9-8245-0f6486b444db.png)

Reply:
Make sense. Thanks for the explanation.